### PR TITLE
Add `tolerable_loss` to `TuningConfig`

### PR DESCRIPTION
**Pull request by Kaihui-intel**
## Type of Change

feature 

## Description
Usage:
``` python
from neural_compressor.torch import RTNConfig, TuningConfig
custom_tune_config = TuningConfig(config_set=[RTNConfig(weight_bits=[4, 6, 8])],  tolerable_loss=-0.01)
```
https://github.com/intel/neural-compressor/discussions/1445#discussioncomment-8152898

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

3x UT

## Dependency Change?

None

### CODE REVIEW

File: neural_compressor/common/base_tuning.py [hunk 1]
```diff
@@ -213,14 +213,21 @@ class TuningConfig:
         config_set: quantization configs. Default value is empty.
         timeout: Tuning timeout (seconds). Default value is 0 which means early stop.
         max_trials: Max tuning times. Default value is 100. Combine with timeout field to decide when to exit.
+        tolerable_loss: The metric loss is absolute, it can be both positive and negative. Default is 0.01.
```
Comment (yiliu30) [inline neural_compressor/common/base_tuning.py:216]: Recommend defining it `relative` to align with the usage in version 2.x.
https://github.com/intel/neural-compressor/blob/9ec22c96993d1a8dc00e0cebca1cbe5527ea8189/neural_compressor/config.py#L502-L510

This argument is crucial for users, so we should provide comprehensive documentation, including the computation formula and examples, to ensure clarity in its usage. Something like(subject to refinement):
```python
# example
tolerable_loss = 0.01
baseline = 100
eval_result_of_nth_trial = 99.1
eval_result_of_nth_trial > baseline * (1 - tolerable_loss ) , reached target.


```

File: neural_compressor/common/base_tuning.py [hunk 2]
```diff
@@ -260,8 +272,17 @@ def get_best_quant_config(self) -> BaseConfig:
         return sorted_trials_records[0].quant_config
 
     def need_stop(self) -> bool:
-        # TODO Support more stop criteria in the next PR, such as `reach accuracy goal`, `timeout`, and so on.
-        return self.trial_cnt >= self.tuning_config.max_trials
+        # TODO Support more stop criteria in the next PR, such as `timeout`, and so on.
```
Comment (yiliu30) [inline neural_compressor/common/base_tuning.py:275]: Please provide comprehensive documentation for the stop logic.

File: test/3x/torch/test_autotune.py [hunk 3]
```diff
@@ -207,6 +207,36 @@ def test_autotune_not_eval_func(self):
...
+
+        def eval_acc_fn(model):
+            res = acc_res_lst.pop(0)
+            return res
+
+        eval_fns = [
+            {"eval_fn": eval_acc_fn, "name": "accuracy"},
+        ]
+
+        custom_tune_config = TuningConfig(config_set=[RTNConfig(weight_bits=[4, 6, 5, 8])], max_trials=6)
```
Comment (yiliu30) [inline test/3x/torch/test_autotune.py:226]: Please include documentation specifying the purpose of each unit test.

File: test/3x/torch/test_autotune.py [hunk 4]
```diff
@@ -207,6 +207,36 @@ def test_autotune_not_eval_func(self):
...
+
+        acc_res_lst = baseline + [0.9] * 2 + [0.99] + [1.01]
+        custom_tune_config = TuningConfig(config_set=[RTNConfig(weight_bits=[4, 6, 5, 8])], tolerable_loss=0.1)
+        best_model = autotune(model=build_simple_torch_model(), tune_config=custom_tune_config, eval_fns=eval_acc_fn)
+        self.assertIsNotNone(best_model)
+
+        acc_res_lst = baseline + [0.9] * 2 + [0.99] + [1.01]
+        custom_tune_config = TuningConfig(config_set=[RTNConfig(weight_bits=[4, 6, 5, 8])], tolerable_loss=-0.01)
+        best_model = autotune(model=build_simple_torch_model(), tune_config=custom_tune_config, eval_fns=eval_acc_fn)
+        self.assertIsNotNone(best_model)
```
Comment (yiliu30) [inline test/3x/torch/test_autotune.py:238]: Add more UTs if possible, for example, the `best_model` is `None`.

File: neural_compressor/common/base_tuning.py (hunk 1 above)
Comment (Kaihui-intel) [reply neural_compressor/common/base_tuning.py]: Sounds good!  When will we write the 3. x document or migration document?https://github.com/intel/neural-compressor/blob/master/docs/source/migration.md

Comment (yiliu30) [reply neural_compressor/common/base_tuning.py]: Let's include this information in the docstring first.

File: neural_compressor/common/base_tuning.py [hunk 5]
```diff
@@ -260,8 +308,23 @@ def get_best_quant_config(self) -> BaseConfig:
...
     def need_stop(self) -> bool:
-        # TODO Support more stop criteria in the next PR, such as `reach accuracy goal`, `timeout`, and so on.
-        return self.trial_cnt >= self.tuning_config.max_trials
+        """Check if need to stop tuning. Either accuracy goal is met, max trials is reached or timeout is reached.
+
+        Returns:
+            bool: True if need to stop, otherwise False.
+        """
+
+        # TODO Support more stop criteria in the next PR, such as `timeout`, and so on.
```
Comment (yiliu30) [inline neural_compressor/common/base_tuning.py:317]:
```suggestion
        # TODO: Support more stop criteria in the next PR, such as `timeout`, and so on.
```

**Review by yiliu30 — APPROVED** [review]

**Review by chensuyue — APPROVED** [review]

**Pull request merged**